Blog: migrations, postgres locks and application downtime (#276)
xataio/mdx-blog: postgres-migrations-locking-and-backoff.mdx

---
title: Schema changes and the Postgres lock queue

> description: 'Learn how schema changes can cause downtime by locking out reads and writes and how migration tools can avoid it by using lock timeouts, along with backoff and retry strategies'

Reviewer: ```suggestion
description: 'Learn how schema changes can cause downtime by locking out reads and writes and how migration tools can avoid it by using lock timeouts, along with backoff and retry strategies.'
```

image:
  src: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/pgroll-internal.jpeg
  alt: pgroll
author: Andrew Farries
authorEmail: andrew@xata.io
date: 06-18-2024
tags: ['open-source', 'postgres', 'schema', 'fpPgroll', 'fpPgzx', 'oss', 'lock', 'migrations']
published: true
slug: migrations-and-exclusive-locks
ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/pgroll-internal.jpeg
---

Schema changes are hard. They're hard because there are multiple different failure modes, some of which require knowledge of how the underlying RDMS system works, especially with regard to object locking.

Broadly speaking, there are two classes of breakage that can occur when applying database migrations:

1. Migrations that make incompatible changes to the schema, breaking client applications.
2. Migrations that lock a database object for an unacceptable amount of time, causing the application to become unavailable as reads and writes start to fail.

The first kind of breakage is perhaps more widely understood and discussed, but the second kind is just as important to consider when running migrations against production systems.

> Today we're going to talk about a specific kind of the second type of breakage: how long running queries together with DDL statements can lock out reads and writes from a table, causing application downtime.

Reviewer: ```suggestion
Today we're going to talk about the second type of breakage: how long running queries together with DDL statements can lock out reads and writes from a table, causing application downtime.
```

## Access exclusive locks

Many Postgres DDL statements attempt to acquire an `ACCESS EXCLUSIVE` lock on the table they are modifying. This lock type is the most aggressive and conflicts with all other lock types. Such DDL statements will block indefinitely until they are able to acquire their lock.

As before, let's simulate a long-running query that acquires an `ACCESS SHARE` lock on the `users` table:

```sql
SELECT *, pg_sleep(30) FROM users
```

Now let's see how a DDL statement blocks while trying to acquire an `ACCESS EXCLUSIVE` lock on the same table:

```sql
ALTER TABLE users ADD COLUMN AGE integer
```

This `ALTER TABLE` statement attempts to acquire an `ACCESS EXCLUSIVE` lock on the `users` table but is unable to do so until the `SELECT` statement completes and releases its `ACCESS SHARE` lock.

So far, so good. What's the problem here? The DDL statement can simply wait patiently until it's able to acquire its `ACCESS EXCLUSIVE` lock, right? The problem is that any other statements that require a lock on the `users` table **are now queued behind this ALTER TABLE statement**, including other `SELECT` statements that only require `ACCESS SHARE` locks.

> This means that the table is effectively blocked for reads and writes until the `ALTER TABLE` statement completes. `SELECT`s and `UPDATE`s will queue up behind it, unable to execute. If there is a long-running query that prevents the `ALTER TABLE` from acquiring the lock, than reads and writes will be blocked for the duration of that query.

Reviewer: ```suggestion
This means that the table is effectively blocked for reads and writes until the `ALTER TABLE` statement completes. `SELECT`s and `UPDATE`s will queue up behind it, unable to execute. If there is a long-running query that prevents the `ALTER TABLE` from acquiring the lock, then reads and writes will be blocked for the duration of that query.
```

## Postgres `lock_timeout` setting

Postgres provides the `lock_timeout` setting to control how long statements should wait to acquire locks before giving up.

By setting a `lock_timeout` on the `ALTER TABLE` statement it's possible to prevent other queries from queueing behind it for an unacceptable length of time:

In one session run:

```sql
SELECT *, pg_sleep(30) FROM users
```

As before, this statement holds an `ACCESS SHARE` lock on the `users` table for 30 seconds per row.

In another session:

```sql
SET lock_timeout TO '1000ms'

ALTER TABLE users ADD COLUMN age INTEGER
```

In a third session:

```sql
SELECT * FROM users
```

After 1 second, the `ALTER TABLE` statement fails with an error like below:

```
canceling statement due to lock timeout
```

Once the `ALTER TABLE` statement has failed, the `SELECT` statement from the third session is unblocked and executes directly.

> DDL statements in migration sessions should always set `lock_timeout` to an appropriate value for the application; values of less than 2 seconds are common. This ensures that reads and writes won't queue behind a blocked DDL statement and cause application downtime.

## Backoff and retry

> With `lock_timeout` set appropriately like this, it's possible to ensure that DDL statements like `ALTER TABLE` don't lock out reads and writes from the affected table. However, when a statement exceeds its `lock_timeout` the statement fails. Postgres does not provide any mechanism by which such statements can be automatically retried. Retrying lock acquisition becomes the responsiblity of the the person or process running the DDL statement.

Reviewer: ```suggestion
With `lock_timeout` set appropriately like this, it's possible to ensure that DDL statements like `ALTER TABLE` don't lock out reads and writes from the affected table. However, when a statement exceeds its `lock_timeout` the statement fails. Postgres does not provide any mechanism by which such statements can be automatically retried. Retrying lock acquisition becomes the responsiblity of the person or process running the DDL statement.
```

> With `pgroll`, our migration tool for Postgres, such lock acquisition failures are automatically retried with an exponential backoff strategy. This means that:

Reviewer: ```suggestion
With `[pgroll](https://github.com/xataio/pgroll)`, our migration tool for Postgres, such lock acquisition failures are automatically retried with an exponential backoff strategy. This means that:
```

Reviewer: First mention of pgroll so wanted to get it in here as well as below.

1. The risk of DDL statements blocking reads and writes is reduced, due to the use of `lock_timeout`.
2. If lock acquisition fails, the DDL is automatically retried and should eventually succeed.

However you make schema changes to your Postgres database, it's important to consider how long running queries together with DDL statements can block reads and writes to a table. Setting `lock_timeout` on DDL statements is a good first step, but it's also important to consider how to handle lock acquisition failures.

# Conclusion

At the top of the post, we identified two common failure modes for database migrations:

1. Schema changes breaking client applications.
2. Unintended downtime due to DDL-induced table locks.

> [pgroll](https://github.com/xataio/pgroll) is a migration tool for Postgres that attempts to address both of these failure modes. It's 'headline feature' is its ability to make two versions, old and new, of a database schema available to client applications. This mitigates the risk of schema changes breaking applications. We've previously blogged about this [here](https://xata.io/blog/pgroll-internals), and the topic is well-covered in the official `pgroll` [docs](https://github.com/xataio/pgroll/tree/main/docs).

Reviewer: ```suggestion
[pgroll](https://github.com/xataio/pgroll) is a migration tool for Postgres that attempts to address both of these failure modes. Its 'headline feature' is its ability to make two versions, old and new, of a database schema available to client applications. This mitigates the risk of schema changes breaking applications. We've previously blogged about this [here](https://xata.io/blog/pgroll-internals), and the topic is well-covered in the official `pgroll` [docs](https://github.com/xataio/pgroll/tree/main/docs).
```
